### PR TITLE
FieldConfig: Set field config defaults based on config registry

### PR DESCRIPTION
**Pull request by dprokop**
Ref #22269

> refactor setFieldConfigDefaults to use registry instead of defaults object

### CODE REVIEW

File: packages/grafana-data/src/field/fieldOverrides.test.ts [hunk 1]
```diff
@@ -276,7 +279,7 @@ describe('setFieldConfigDefaults', () => {
     expect(dsFieldConfig).toMatchInlineSnapshot(`
       Object {
         "custom": Object {
-          "property1": 20,
```
Comment (dprokop) [inline packages/grafana-data/src/field/fieldOverrides.test.ts:279]: In my previous PR I've updated this test to this wrong value

File: packages/grafana-data/src/field/fieldOverrides.ts [hunk 2]
```diff
@@ -254,20 +247,21 @@ export function setFieldConfigDefaults(config: FieldConfig, defaults: FieldConfi
 const processFieldConfigValue = (
   destination: Record<string, any>, // it's mutable
   source: Record<string, any>,
-  key: string,
+  fieldConfigProperty: FieldPropertyEditorItem,
```
Comment (torkelo) [inline packages/grafana-data/src/field/fieldOverrides.ts:250]: think FieldPropertyEditorItem should be renamed to FieldConfigPropertyItem, as it's more than just an editor, is really a descriptor for a Field Config Property. but very important, but something I was thinking when working on the previous PR

File: packages/grafana-data/src/field/fieldOverrides.ts [hunk 3]
```diff
@@ -226,26 +229,16 @@ function setDynamicConfigValue(config: FieldConfig, value: DynamicConfigValue, c
...
-
-        const customKeys = Object.keys(defaults.custom!);
-        for (const customKey of customKeys) {
-          processFieldConfigValue(config.custom!, defaults.custom!, `custom.${customKey}`, context);
-        }
-      } else {
-        // when config from ds exists for a given field -> use it
-        processFieldConfigValue(config, defaults, key, context);
-      }
+  for (const fieldConfigProperty of context.fieldConfigRegistry.list()) {
```
Comment (torkelo) [inline packages/grafana-data/src/field/fieldOverrides.ts:232]: Idea (Not related to this PR):  making the Registry class implement [Symbol.iterator](){ woudl be nice , then we would not need to call list() and we could for each over it directly.

**Review by torkelo — APPROVED** [review]
Looks good!

**Review by ryantxu — APPROVED** [review]
looks great

**Comment by torkelo** [comment]
One thing that the other PR leads to is this
image

The new custom options as saved with a nested id as well as a isCustom:true property.  Need to decide on what to do there, as using both seem a bit too much,

**Comment by dprokop** [comment]
@torkelo simplified the persistence model (only `id` is persisted for overrides). Also added nested paths support for overrides

**Comment by dprokop** [comment]
So, this is ready for final review

File: packages/grafana-data/src/field/fieldOverrides.ts [hunk 4]
```diff
@@ -209,19 +208,19 @@ function setDynamicConfigValue(config: FieldConfig, value: DynamicConfigValue, c
   const remove = val === undefined || val === null;
 
   if (remove) {
-    if (value.isCustom && config.custom) {
-      delete config.custom[item.path];
+    if (item.isCustom && config.custom) {
+      delete config.custom[item.path.split('.')[0]];
```
Comment (torkelo) [inline packages/grafana-data/src/field/fieldOverrides.ts:212]: not sure I understand this, introducing path was to avoid splitting id

Comment (torkelo) [reply packages/grafana-data/src/field/fieldOverrides.ts:212]: and it looks strange, is custom.nestedProp.value is undefined, wht is custom.nestedProp deleted?

File: packages/grafana-ui/src/components/Table/Table.story.tsx [hunk 5]
```diff
@@ -105,8 +105,8 @@ export const BarGaugeCell = () => {
     {
       matcher: { id: FieldMatcherID.byName, options: 'Progress' },
       properties: [
-        { id: 'width', value: '200', isCustom: true },
-        { id: 'displayMode', value: 'gradient-gauge', isCustom: true },
+        { id: 'width', value: '200' },
+        { id: 'displayMode', value: 'gradient-gauge' },
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/Table.story.tsx:109]: should the id's here not be "custom.width" if we remove isCustom?

File: packages/grafana-data/src/field/fieldOverrides.ts (hunk 4 above)
Comment (dprokop) [reply packages/grafana-data/src/field/fieldOverrides.ts:212]: but you can't `delete` based on path (if nested).

File: packages/grafana-ui/src/components/Table/Table.story.tsx (hunk 5 above)
Comment (dprokop) [reply packages/grafana-ui/src/components/Table/Table.story.tsx:109]: good catch

File: packages/grafana-data/src/field/fieldOverrides.test.ts [hunk 6]
```diff
@@ -35,8 +36,16 @@ const property2 = {
   shouldApply: () => true,
 } as any;
 
+const property3 = {
+  id: 'custom.property3', // Match field properties
```
Comment (torkelo) [inline packages/grafana-data/src/field/fieldOverrides.test.ts:40]: would this not be custom.property3.nested ?

File: packages/grafana-data/src/field/fieldOverrides.test.ts [hunk 7]
```diff
@@ -286,3 +295,118 @@ describe('setFieldConfigDefaults', () => {
...
+      custom: {
+        property3: {
+          nested: 1,
+        },
+      },
+    };
+    setDynamicConfigValue(
+      config,
+      {
+        id: 'custom.property3',
```
Comment (torkelo) [inline packages/grafana-data/src/field/fieldOverrides.test.ts:358]: custom.property3.nested

File: packages/grafana-data/src/field/fieldOverrides.ts (hunk 4 above)
Comment (torkelo) [reply packages/grafana-data/src/field/fieldOverrides.ts:212]: then I think we should not support nested props, or I don't know, this looks like it does not work either.

So this remove clause, do we need this?  what are we removing here? where is the thing we are removing coming from? are we removing defaults?

Comment (torkelo) [reply packages/grafana-data/src/field/fieldOverrides.ts:212]: not sure we need the remove/delete handling here: 

only when we modify the persisted model: 
https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/components/PanelEditor/FieldConfigEditor.tsx#L125 

Wonder if we should share some code between this function and the function used by the editor to remove & set default values

Comment (dprokop) [reply packages/grafana-data/src/field/fieldOverrides.ts:212]: yeah, we probably could, i'll followup on this ltr

**Pull request merged**